### PR TITLE
Accelerate lister::rmdir

### LINKED ISSUE #10484: lister::rmdir is extremely slow
Reproduced using a unit test that creates thousands of entries in a shallow tree,
lister::rmdir demonstrated throughput of less than 10 entries per second.
It turns out that calling remove_file for each entry is a bad idea since it
triggers full path lookup in the kernel, and `lister::rmdir` should use `unlinkat`
instead, since it already has the listed directory file opened.

### PR DESCRIPTION
**Pull request by bhalevy**
Calling remove_file for each entry is slow
since it requires to do a full path lookup for each entry.

This series adds a unit test to exercise lister::tmdir
and adds lister's own implementation
of synchronous list_directory (allowing yielding
in between buffers), and then uses `unlinkat` to
unlink each file, reusing the open directory
file handle.

Fixes #10484

### CODE REVIEW

**Comment by bhalevy** [comment]
v2:
- update description and git commit log

File: utils/lister.cc [hunk 1]
```diff
@@ -61,23 +65,112 @@ future<> lister::scan_dir(fs::path dir, lister::dir_entry_types type, lister::sh
...
+        if (name == "." || name == "..") {
+            continue;
+        }
+        ret.emplace_back(directory_entry{std::move(name), type});
+    }
+
+    return ret;
+}
+
 future<> lister::rmdir(fs::path dir) {
```
Comment (xemul) [inline utils/lister.cc:130]: Patch title says it's `mkdir` re-implementation

File: utils/lister.cc [hunk 2]
```diff
@@ -61,23 +65,112 @@ future<> lister::scan_dir(fs::path dir, lister::dir_entry_types type, lister::sh
...
+                    if (errno != ENOENT) {
+                        co_await coroutine::return_exception(std::system_error(std::error_code(errno, std::system_category()), (dir / de.name).native()));
+                    }
+                }
+                co_await coroutine::maybe_yield();
+            }
+        }
+        for (auto& de : ents) {
+            if (de.type == directory_entry_type::directory) {
+                co_await rmdir(dir / de.name);
```
Comment (xemul) [inline utils/lister.cc:164]: You can use `unlinkat(..., AT_REMOVEDIR)` to remove a subdirectory in the -at manner.

File: utils/lister.cc (hunk 1 above)
Comment (bhalevy) [reply utils/lister.cc:130]: fixed

File: utils/lister.cc (hunk 2 above)
Comment (xemul) [reply utils/lister.cc:164]: I intented to put this comment few lines below, where the `auto res = ::rmdir(dir.native().c_str());` is called, but I agree with you -- with shallow recursion this benefit can be achived later on demand.

**Comment by avikivity** [comment]
> Calling remove_file for each entry is slow since it requires to do a full path lookup for each entry.
> 

I can't believe this takes anything noticable. It should be less than a microsecond.

> This series adds a unit test to exercise lister::tmdir and adds lister's own implementation of synchronous list_directory (allowing yielding in between buffers), and then uses `unlinkat` to unlink each file, reusing the open directory file handle.
> 

unlinkat()? and what if it blocks?

We don't call file system calls in the reactor thread.

> Fixes #10484

File: utils/lister.cc [hunk 3]
```diff
@@ -61,23 +65,112 @@ future<> lister::scan_dir(fs::path dir, lister::dir_entry_types type, lister::sh
...
+
+    while (true) {
+        auto ents = list_directory(dirfd);
+        if (ents.empty()) {
+            break;
+        }
+        for (auto& de : ents) {
+            if (de.type != directory_entry_type::directory) {
+                llogger.debug("rmdir: unlinking {} from {}", de.name, dir);
+                auto res = ::unlinkat(dirfd, de.name.c_str(), 0);
```
Comment (avikivity) [inline utils/lister.cc:153]: A big no-no.

Comment (avikivity) [reply utils/lister.cc:153]: I think you misdiagnosed the problem.

```c++
future<> lister::rmdir(fs::path dir) {
    // first, kill the contents of the directory
    return lister::scan_dir(dir, {}, show_hidden::yes, [] (fs::path parent_dir, directory_entry de) mutable {
        fs::path current_entry_path(parent_dir / de.name.c_str());

        if (de.type.value() == directory_entry_type::directory) {
            return rmdir(std::move(current_entry_path));
        } else {
            return remove_file(current_entry_path.native());
        }
```

We have to wait for each remove_file() before proceeding. If there is a background load (like compaction), this takes task_quota_ms. So we remove files at a rate of 2000 files/sec/shard.

If this is too slow, the solution is to remove files in parallel (only wait for removal to complete before removing the parent directory).

**Comment by bhalevy** [comment]
retracting

**Pull request closed without merging**